docs/development.rst

```diff
@@ -2,9 +2,177 @@
     :start-after: sec-begin-development
     :end-before: sec-end-development
 
+.. _writing-adapters:
+
 Writing a model adapter
 =======================
 
+Creating an :class:`~openscm.adapter.Adapter` subclass
+******************************************************
+
+- Create your adapter source file in ``openscm/adapters/``, e.g.
+  ``myadapter.py``
+
+- If your adapter needs additional dependencies add them to the
+  ``requirements_extras`` dictionary in ``setup.py`` (see comment
+  there).
+
+- Subclass the :class:`openscm.adapter.Adapter` class:
+
+  .. code:: python
+
+      from ..adapter import Adapter
+
+      class MyAdapter(Adapter):
+
+- Implement the relevant methods (or just do ``pass`` if you do not
+  need to do anything in the particular method). An adapter should not
+  use parts of OpenSCM other than ``ParameterSet``.
+
+  - The :func:`~openscm.adapter.Adapter._initialize_model` method
+    initializes the adapter and is called only once. It should set the
+    default values of mandatory *model-specific* (not :ref:`standard
+    OpenSCM parameters <standard-parameters>`!) parameters in the
+    :class:`~openscm.core.ParameterSet`. Their :ref:`hierarchical
+    names <parameter-hierarchy>` should start with the model/adapter
+    name.
+
+    .. code:: python
+
+        def _initialize_model(self) -> None:
```

Comment:
This needs to take arguments as how some models initialise varies depending on how you want to use them (see e.g. the `root_dir` argument to the `MAGICCBase` class in Pymagicc). Unless you want `__init__` to take arguments instead?

Reply:
> Why is this private? Cause it should only ever be called by `Core` and not by the user directly?

I wanted to have everything the (dev) user implements be protected so we can later wrap these methods in the `Adapter` base class (e.g., as for `step`). The (end) user never calls the adapter, only ``Core``.

> Somewhere in there, the user needs to be able to pass arguments to an initialisation step as how some models initialise varies depending on how you want to use them (see e.g. the `root_dir` argument to the `MAGICCBase` class in Pymagicc). Those arguments could be passed here or in `__init__` instead?

Hmm, maybe I here cling to a language-independent interface too much. For that, we would require all such parameters be set in ``ParameterSet`` (the “model-specific parameters” in the doc). Maybe then postpone the ``_initialize_model`` till the first initialization of a run?

Reply:
> maybe I here cling to a language-independent interface too much

Maybe, no idea. 

> Maybe then postpone the `_initialize_model` till the first initialization of a run?

Can do that too. 

> For that, we would require all such parameters be set in `ParameterSet`

Ok then we would need a new parameter type, string (or I would actually lean to generic where people can put whatever random crap they need to get their model to run) to pass stuff like `run_dir` (which could be e.g. a path pointing to where your executable is).

Reply:
see #124